src/lib/server/postgres_protocol_handler.cpp

```diff
@@ -0,0 +1,310 @@
+}
+
+template <typename SocketType>
+std::string PostgresProtocolHandler<SocketType>::read_query_packet() {
+  const auto query_length = _read_buffer.template get_value<uint32_t>() - LENGTH_FIELD_SIZE;
+  return _read_buffer.get_string(query_length);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::_ssl_deny() {
+  // The SSL deny packet has a special format. It does not have a field indicating the packet size.
+  _write_buffer.template put_value(PostgresMessageType::SslNo);
+  _write_buffer.flush();
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::read_startup_packet_body(const uint32_t size) {
+  // As of now, we don't do anything with the startup packet body. It contains authentication data and the
+  // database name the user desires to connect to. Hence, we read this information from the network device as
+  // one large string and throw it away.
+  _read_buffer.get_string(size, false);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_authentication() {
+  _write_buffer.template put_value(PostgresMessageType::AuthenticationRequest);
+  // Since we don't have any authentication mechanism, authentication is always successful
+  constexpr uint32_t authentication_successful = 0;
+  _write_buffer.template put_value<uint32_t>(sizeof(LENGTH_FIELD_SIZE) + sizeof(authentication_successful));
+  _write_buffer.template put_value<uint32_t>(authentication_successful);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_parameter(const std::string& key, const std::string& value) {
+  const auto packet_size = sizeof(LENGTH_FIELD_SIZE) + key.size() + value.size() + 2u /* null terminator */;
+  _write_buffer.template put_value(PostgresMessageType::ParameterStatus);
+  _write_buffer.template put_value<uint32_t>(static_cast<uint32_t>(packet_size));
+  _write_buffer.put_string(key);
+  _write_buffer.put_string(value);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_ready_for_query() {
+  _write_buffer.template put_value(PostgresMessageType::ReadyForQuery);
+  _write_buffer.template put_value<uint32_t>(sizeof(LENGTH_FIELD_SIZE) + sizeof(TransactionStatusIndicator::Idle));
+  _write_buffer.template put_value(TransactionStatusIndicator::Idle);
+  _write_buffer.flush();
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_command_complete(const std::string& command_complete_message) {
+  const auto packet_size = sizeof(LENGTH_FIELD_SIZE) + command_complete_message.size() + 1u /* null terminator */;
+  _write_buffer.template put_value(PostgresMessageType::CommandComplete);
+  _write_buffer.template put_value<uint32_t>(static_cast<uint32_t>(packet_size));
+  _write_buffer.put_string(command_complete_message);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::set_row_description_header(const uint32_t total_column_name_length,
```

Comment:
Both this and `send_row_description` write to the output buffer, but one is called "set" and the other one "send"